icu4j/main/classes/core/src/com/ibm/icu/message2/Mf2DataModelFormatter.java

```diff
@@ -0,0 +1,277 @@
+                .setDefaultFormatterNameForType(Integer.class, "number")
+                .setDefaultFormatterNameForType(Double.class, "number")
+                .setDefaultFormatterNameForType(Number.class, "number")
+                .setDefaultFormatterNameForType(CurrencyAmount.class, "number")
+
+                // Format that returns "to string"
+                .setFormatter("identity", new IdentityFormatterFactory())
+                .setDefaultFormatterNameForType(String.class, "identity")
+                .setDefaultFormatterNameForType(CharSequence.class, "identity")
+
+                // Register the standard selectors
+                .setSelector("plural", new PluralSelectorFactory("cardinal"))
+                .setSelector("selectordinal", new PluralSelectorFactory("ordinal"))
+                .setSelector("select", new TextSelectorFactory())
+                .setSelector("gender", new TextSelectorFactory())
+
+                .build();
+    }
+
+    private static Map<String, Object> mf2OptToFixedOptions(Map<String, Value> options) {
+        Map<String, Object> result = new HashMap<>();
+        for (Entry<String, Value> option : options.entrySet()) {
+            Value value = option.getValue();
+            if (value.isLiteral()) {
+                result.put(option.getKey(), value.getLiteral());
+            }
+        }
+        return result;
+    }
+
+    private Map<String, Object> mf2OptToVariableOptions(Map<String, Value> options, Map<String, Object> arguments) {
+        Map<String, Object> result = new HashMap<>();
+        for (Entry<String, Value> option : options.entrySet()) {
+            Value value = option.getValue();
+            if (value.isVariable()) {
+                result.put(option.getKey(), variableToObjectEx(value, arguments));
+            }
+        }
+        return result;
+    }
+
+    FormatterFactory getFormattingFunctionFactoryByName(Object toFormat, String functionName) {
+        // Get a function name from the type of the object to format
+        if (functionName == null || functionName.isEmpty()) {
+            if (toFormat == null) {
+                // The object to format is null, and no function provided.
+                return null;
+            }
+            Class<?> clazz = toFormat.getClass();
+            functionName = standardFunctionRegistry.getDefaultFormatterNameForType(clazz);
+            if (functionName == null) {
+                functionName = customFunctionRegistry.getDefaultFormatterNameForType(clazz);
+            }
+            if (functionName == null) {
+                throw new IllegalArgumentException("Object to format without a function, and unknown type: "
+                        + toFormat.getClass().getName());
+            }
+        }
+
+        FormatterFactory func = standardFunctionRegistry.getFormatter(functionName);
+        if (func == null) {
+            func = customFunctionRegistry.getFormatter(functionName);
+        }
```

Comment:
ditto which one to look in first? probably same elsewhere for selectors etc.

Reply:
Same answer.
I don't feel strongly one way or another, and I can see pros and cons both ways.